commands.go

```diff
@@ -92,14 +94,152 @@ func wrapCommand(handler func(*WrappedCommandEvent)) func(*commands.Event) {
 	}
 }
 
+type StateHandler struct {
+	Func func(*WrappedCommandEvent)
+	Name string
+}
+
+func (sh *StateHandler) Run(ce *commands.Event) {
+	wrapCommand(sh.Func)(ce)
+}
+
+func (sh *StateHandler) GetName() string {
+	return sh.Name
+}
+
 var (
 	HelpSectionConnectionManagement = commands.HelpSection{Name: "Connection management", Order: 11}
 	HelpSectionCreatingPortals      = commands.HelpSection{Name: "Creating portals", Order: 15}
 	HelpSectionPortalManagement     = commands.HelpSection{Name: "Portal management", Order: 20}
 	HelpSectionInvites              = commands.HelpSection{Name: "Group invites", Order: 25}
 	HelpSectionMiscellaneous        = commands.HelpSection{Name: "Miscellaneous", Order: 30}
+
+	roomArgHelpMd = " [<_Matrix room ID_> | --here]"
+	roomArgHelp   = " [<Matrix room ID> | --here]"
 )
 
+func getBridgeRoomID(ce *WrappedCommandEvent, argIndex int) (roomID id.RoomID, ok bool) {
+	if len(ce.Args) <= argIndex {
+		ok = true
+		return
+	}
+	roomArg := ce.Args[argIndex]
+	if roomArg == "--here" {
+		roomID = ce.RoomID
+	} else if strings.HasPrefix(roomArg, "!") {
+		roomID = id.RoomID(roomArg)
+	} else if strings.HasPrefix(roomArg, "#") {
+		resp, err := ce.MainIntent().ResolveAlias(id.RoomAlias(roomArg))
+		if err != nil {
+			ce.Log.Errorln("Failed to resolve room alias %s to a room ID: %v", roomArg, err)
+			ce.Reply("Unable to find a room with the provided alias")
+			return
+		} else {
+			roomID = resp.RoomID
+		}
+	}
+	if roomID == "" {
+		ce.Reply("Invalid room ID")
+		return
+	}
+
+	var thatThisSuffix string
+	if roomID == ce.RoomID {
+		thatThisSuffix = "is"
+	} else {
+		thatThisSuffix = "at"
+	}
+
+	portal := ce.Bridge.GetPortalByMXID(roomID)
+	if portal != nil {
+		ce.Reply("Th%s room is already a portal room.", thatThisSuffix)
+	} else if !userHasPowerLevel(roomID, ce.MainIntent(), ce.User, "bridge") {
+		ce.Reply("You do not have the permissions to bridge th%s room.", thatThisSuffix)
+	} else {
+		ok = true
+	}
+	return
+}
+
+func getInitialState(intent *appservice.IntentAPI, roomID id.RoomID) (
+	name string,
+	topic string,
+	levels *event.PowerLevelsEventContent,
+	encrypted bool,
+) {
+	state, err := intent.State(roomID)
+	if err == nil {
+		for _, events := range state {
+			for _, evt := range events {
+				switch evt.Type {
+				case event.StateRoomName:
+					name = evt.Content.AsRoomName().Name
+				case event.StateTopic:
+					topic = evt.Content.AsTopic().Topic
+				case event.StatePowerLevels:
+					levels = evt.Content.AsPowerLevels()
+				case event.StateEncryption:
+					encrypted = true
```

Comment:
/me googles if Go does not have the concept of `break` in switch statements.

Reply:
Go has `fallthrough` instead of `break`. Nice! :+1:

Reply:
It doesn't: "[the break statement that is needed at the end of each [switch] case in [other] languages is provided automatically in Go.](https://go.dev/tour/flowcontrol/9)"